GitUI/Avatars/CustomAvatarProvider.cs

```diff
@@ -0,0 +1,102 @@
+    /// are queried in order until an avatar image is found.
+    /// </summary>
+    /// <remarks>
+    /// Details about the usage and syntax of the custom avatar provider can be found in the gitextensions wiki.
+    /// </remarks>
+    public sealed partial class CustomAvatarProvider : IAvatarProvider
+    {
+        private readonly IAvatarDownloader _downloader;
+        private readonly IAvatarProvider[] _subProvider;
+
+        private CustomAvatarProvider(IAvatarDownloader downloader, IAvatarProvider[] subProvider)
+        {
+            _downloader = downloader ?? throw new ArgumentNullException(nameof(downloader));
+            _subProvider = subProvider ?? throw new ArgumentNullException(nameof(subProvider));
+        }
+
+        /// <inheritdoc/>
+        public async Task<Image> GetAvatarAsync([NotNull] string email, string name, int imageSize)
+        {
+            var templateData = new UriTemplateData(email, name, imageSize);
+
+            foreach (var provider in _subProvider)
+            {
+                var avatar = provider switch
+                {
+                    UriTemplateResolver r => await _downloader.DownloadImageAsync(r.ResolveTemplate(templateData)),
+                    _ => await provider.GetAvatarAsync(email, name, imageSize),
+                };
+
+                if (avatar != null)
+                {
+                    return avatar;
+                }
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Parses a custom avatar template string and creates an <see cref="IAvatarProvider"/> from it.
+        /// </summary>
+        /// <param name="customProviderTemplates">The custom avatar provider template.</param>
+        /// <param name="downloader">The downloader that is used to download avatar images.</param>
+        /// <returns>Returns the <see cref="IAvatarProvider"/> described by the template.</returns>
+        public static IAvatarProvider ParseTemplateString(string customProviderTemplates, [NotNull] IAvatarDownloader downloader)
+        {
+            if (downloader is null)
+            {
+                throw new ArgumentNullException(nameof(downloader));
+            }
+
+            var providerParts = customProviderTemplates
+                .Split(';')
+                .Select(p => p.Trim())
+                .Select(p => FromTemplateSegment(downloader, p))
+                .Where(p => p != null)
+                .ToArray();
+
+            // We can't use the chain provider here, because some returned providers (namely UriTemplateResolvers)
+            // don't actually fulfill the interface contract of IAvatarProvider. UriTemplateResolver is a special
+            // case that exists to prevent variables like hashes from being evaluated/calculated multiple times.
+
+            return new CustomAvatarProvider(downloader, providerParts);
+        }
```

Comment:
Please first split parameters then create CustomAvatarProvider(downloader, providerParts) with UriTemplateResolver if needed and not use providerParts as parameter. I think this help refactor code in FromTemplateSegment.
Like this.

```
var templates = customProviderTemplates
    .Split(new string[] { ";" }, StringSplitOptions.RemoveEmptyEntries)
    .Select(x => x.Trim())
    .Where(x => x != string.Empty)
    .ToList();

foreach (var template in templates)
{
    if (template.StartsWith("<") && template.EndsWith(">"))
    {
        // do changed FromTemplateSegment
        var avatarProvider = FromTemplateSegment(downloader, template)

        return new CustomAvatarProvider(downloader, avatarProvider);
    }

    return new CustomAvatarProvider(downloader, new UriTemplateResolver(template));
}
```

Reply:
And please rename method, this is factory method but named as "Parse".

Reply:
But it is parsing the template string. By the same argument `DateTime.Parse` should not be called Parse because it's the factory of a `DateTime`

Reply:
DateTime.Parse => DateTime.
AvatarProvider.Parse => AvatarProvider.

but

CustomAvatarProvider.ParseTemplateString => IAvatarProvider.

Reply:
True, so if I'd change the return type to `CustomAvatarProvider` than `Parse` as a name would be fine?

Reply:
😄